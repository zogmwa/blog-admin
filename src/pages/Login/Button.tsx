import React from 'react';
import { ButtonProps } from 'antd/lib/button';
import { Button } from 'antd';
import { ApolloConsumer } from 'react-apollo';
import ApolloClient from 'apollo-client';

type LoginButtonProps = {
  login: (e: React.MouseEvent, client: ApolloClient<any>) => void;
} & ButtonProps;

export const LoginButton = ({ onClick, login, ...props }: LoginButtonProps) => (
  <ApolloConsumer>
    {client => (
      <Button {...props} onClick={(e: React.MouseEvent) => login(e, client)}>
        Login
      </Button>
    )}
  </ApolloConsumer>
);
