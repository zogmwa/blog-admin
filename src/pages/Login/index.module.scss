.container {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100vh;
  background: url(https://static.jkchao.cn/main_bg.jpg) no-repeat 20% 20%;
  background-size: cover;
}

.form {
  width: 250px;
}

.form-button {
  width: 100%;
}
