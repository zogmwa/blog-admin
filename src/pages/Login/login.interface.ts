import { FormComponentProps } from 'antd/lib/form';
import { RouteComponentProps } from 'react-router';

export interface Variables {
  username: string;
  password: string;
}

export interface LoginState {
  loading: boolean;
}

export interface Response {
  login: {
    lifeTime: number;
    token: string;
  };
}

export interface UserFormProps
  extends FormComponentProps,
    RouteComponentProps,
    Variables {}
