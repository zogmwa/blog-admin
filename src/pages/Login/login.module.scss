.form {
  max-width: 300px;
}

.form-button {
  width: 100%;
}
