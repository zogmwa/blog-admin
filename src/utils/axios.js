import axios from 'axios'
import * as config from '../config.js'
// import router from '../router'

const service = axios.create({
  baseURL: config.API_ROOT
})

// 拦截器
service.interceptors.request.use(config => {
  if (window.localStorage.getItem('TOKEN')) {
    config.headers.Authorization = window.localStorage.getItem('TOKEN')
  }
  return config
}, error => {
  return Promise.reject(error)
})

service.interceptors.response.use(response => {
  // if (response.data.code === 2) router.push('/login')
  return response
}, error => {
  return Promise.reject(error)
})

export default service
