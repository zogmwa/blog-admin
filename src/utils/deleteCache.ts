import { DataProxy } from 'apollo-cache';

// https://github.com/apollographql/apollo-feature-requests/issues/4#issuecomment-431119231
export const deleteCache = (cache: DataProxy, key: RegExp) => {
  // @ts-ignore
  Object.keys(cache.data.data).forEach(key => {
    // @ts-ignore
    key.match(key) && cache.data.delete(key);
  });
};
