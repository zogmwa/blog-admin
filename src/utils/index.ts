export { urlToList } from './pathToList';
export { checkLogin } from './isLogin';
export { deleteCache } from './deleteCache';
