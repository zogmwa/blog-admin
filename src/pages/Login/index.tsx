import { Form, Icon, Input, notification } from 'antd';
import ApolloClient from 'apollo-client';
import React from 'react';

import { LoginButton } from './Button';
import styles from './index.module.scss';
import {
  LoginState,
  Response,
  UserFormProps,
  Variables
} from './index.interface';
import { LOGIN } from './index.query';

class Login extends React.PureComponent<UserFormProps, LoginState> {
  state = {
    loading: false
  };

  handleSubmit = (e: React.MouseEvent, client: ApolloClient<any>) => {
    this.props.form.validateFields(async (err, values) => {
      if (!err) {
        this.setState({
          loading: true
        });

        try {
          const {
            data: { login }
          } = await client.query<Response, Variables>({
            query: LOGIN,
            variables: values,
            errorPolicy: 'all'
          });

          this.setState({
            loading: false
          });

          if (login) {
            window.localStorage.setItem('TOKEN', JSON.stringify(login));
            const path = this.props.location.state.from.pathname;
            this.props.history.push(path || '/dashboard');
          }
        } catch (error) {
          this.setState({
            loading: false
          });
        }
      }
    });
  };

  render() {
    const { getFieldDecorator } = this.props.form;
    return (
      <div className={styles.container}>
        <Form className={styles.form}>
          <Form.Item hasFeedback>
            {getFieldDecorator('username', {
              rules: [
                { required: true, message: '请输入姓名' },
                { min: 6, message: '最小长度为 6 位' }
              ]
            })(
              <Input
                prefix={
                  <Icon type="user" style={{ color: 'rgba(0,0,0,.25)' }} />
                }
                placeholder="Username"
              />
            )}
          </Form.Item>
          <Form.Item>
            {getFieldDecorator('password', {
              rules: [
                { required: true, message: 'Please input your Password!' }
              ]
            })(
              <Input
                prefix={
                  <Icon type="lock" style={{ color: 'rgba(0,0,0,.25)' }} />
                }
                type="password"
                placeholder="Password"
              />
            )}
          </Form.Item>
          <Form.Item>
            <LoginButton
              type="primary"
              htmlType="submit"
              className={styles['form-button']}
              login={this.handleSubmit}
              loading={this.state.loading}
            >
              Log in
            </LoginButton>
          </Form.Item>
        </Form>
      </div>
    );
  }
}

export default Form.create()(Login);
