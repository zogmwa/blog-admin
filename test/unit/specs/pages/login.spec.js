import Vue from 'vue'
import Vuex from 'vuex'
import ElementUI from 'element-ui'
import { mount, createLocalVue } from 'vue-test-utils'

import Login from '@/pages/login'
import { destroyVm } from '../../utils/utils'

Vue.use(ElementUI)

// 独立 Vue 构造函数
const localVue = createLocalVue()

localVue.use(Vuex)

describe('login.vue', () => {
  let wrapper
  let actions
  let state
  let store

  afterEach(() => {
    destroyVm(wrapper.vm)
  })

  beforeEach(() => {
    state = {
      loading: true
    }
    actions = {
      login: jest.fn()
    }
    store = new Vuex.Store({
      state,
      actions
    })
  })

  it('Renders state.loading', () => {
    wrapper = mount(Login, { store, localVue })
    const btn = wrapper.find('.el-button')
    expect(btn.classes()).not.toContain('is-disabled')
  })

  it('Username and password should in page', async () => {
    wrapper = mount(Login, {
      store,
      localVue
    })
    const input = wrapper.findAll('.el-input')
    expect(input.length).toBe(2)
    wrapper.setData({
      form: {
        username: 'test username',
        password: 'test password'
      }
    })
    expect(wrapper.vm.form.username).toBe('test username')
    expect(wrapper.vm.form.password).toBe('test password')
  })

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
