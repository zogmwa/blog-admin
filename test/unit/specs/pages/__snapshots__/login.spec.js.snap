// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login.vue has the expected html structure 1`] = `
<div
  class="login"
>
  <div
    class="background-canvas"
    id="particles-background"
  />
   
  <div
    class="content"
  >
    <div
      class="title"
    >
      管理登录
    </div>
     
    <form
      class="el-form"
    >
      <div
        class="el-form-item is-required"
      >
        <!---->
        <div
          class="el-form-item__content"
        >
          <div
            class="el-input"
          >
            <!---->
            <input
              autocomplete="off"
              class="el-input__inner"
              maxlength="40"
              placeholder="账号"
              rows="2"
              type="text"
              validateevent="true"
            />
            <!---->
            <!---->
            <!---->
          </div>
          <!---->
        </div>
      </div>
       
      <div
        class="el-form-item is-required"
      >
        <!---->
        <div
          class="el-form-item__content"
        >
          <div
            class="el-input"
          >
            <!---->
            <input
              autocomplete="off"
              class="el-input__inner"
              maxlength="40"
              placeholder="密码"
              rows="2"
              type="password"
              validateevent="true"
            />
            <!---->
            <!---->
            <!---->
          </div>
          <!---->
        </div>
      </div>
       
      <div
        class="el-form-item"
      >
        <!---->
        <div
          class="el-form-item__content"
        >
          <button
            class="el-button el-button--default"
            type="button"
          >
            <!---->
            <!---->
            <span>
              Submit
            </span>
          </button>
          <!---->
        </div>
      </div>
    </form>
  </div>
</div>
`;
